### DISCUSSION TITLE
Transfer learning using custom training data for 3.0.0 training

### DISCUSSION THREAD

**Discussion by rszabo15** [explosion/spaCy#7555] [Help: Coding & Implementations] [answered]
Hi - 
I am interested in doing transfer leaning using the 'en_core_web_trf' as a basis and updating the model with custom training data using the configuration file setup. 
Since I am transferring my model from version 2, I have training data in the following format
[("Walmart is a leading e-commerce company", {"entities": [(0, 7, "ORG")]}),
              ("I reached Chennai yesterday.", {"entities": [(19, 28, "GPE")]})]

A few questions that I couldn't figure out from the documentation:
1. Can I continue to use this format for the training data?
2. Can you give an example of how to save the training data in the right file format and call this training file in the cli 'train' function?
3. How do I setup the config file to ensure that I am using en_core_web_trf as a basis for transfer learning?
4. Any additional commands/steps needed to do the transfer learning

Thank you

**Comment by polm** [comment, marked as answer]
> Can I continue to use this format for the training data?

Yes, basically. You may need to make slight changes depending on how you feed the examples into the models. See the [training data section](https://spacy.io/api/data-formats#dict-input) in the docs for more details on that. 

> How do I setup the config file to ensure that I am using en_core_web_trf as a basis for transfer learning?

There's a [quickstart section](https://spacy.io/usage/training#quickstart) that lets you create a config for training a model from a Transformer. Did you have any issues using that?

**Reply by rszabo15** [reply]
For the training data, I am trying to convert a json list of Example data to the .spacy format using
`python -m spacy convert ./data.json ./output.spacy`
Where data.json is the training data Example objects. However, I'm getting the error 'Invalid value for '[OUTPUT_DIR]': path ./output.spacy does not exist.
Can you explain the input and output file format necessary?

For the quickstart section, I used the basic configuration that is provided, but I am not certain where/how to (or if I need to) specify that I want to train based off of the en_core_web_trf model and where/how to put the additional training data that I want to use. 
Can you please clarify?

**Reply by polm** [reply]
> However, I'm getting the error 'Invalid value for '[OUTPUT_DIR]': path ./output.spacy does not exist.

Ah, that's an issue with the docs - the output path should be a directory, not a file. So you can just use `.`, or you can make an `output` directory. 

As a note, I don't think the convert command works with that format - the old JSON format was a specific different format. This is an example of how you would load your data:

```
TRAINING_DATA = ... # a list of your training data instances
examples = []
for text, gold_dict in TRAINING_DATA:
    example = Example.from_dict(nlp(text), gold_dict)
    examples.append(example)
```

> For the quickstart section, I used the basic configuration that is provided ...

In the quickstart config you'll see a reference to `roberta-base` - that Transformer model will be used as a base for training, so you get transfer learning by default. If you want to fine-tune on the full `en_core_web_trf` model, you can do that on a per-component basis. If you look at [the docs on training components](https://spacy.io/usage/training#config-components) you'll see that you can use a section like this to initialize a component from a pretrained model:

```
[components.parser]
source = "en_core_web_sm"
```

also keep this point in mind:

> By default, spaCy will update all components that can be updated. 

So if you load a pretrained component, by default it will be fine-tuned (used for transfer learning).

**Reply by rszabo15** [reply]
> This is an example of how you would load your data:

When I copy and paste the contents for `examples` list to a json file and try to convert the json to spacy using 
`python -m spacy convert ./train.json .`
I get a a `KeyError 'paragraphs'` 
Is there anything else that I need to add to the examples data output before conversion?

**Reply by polm** [reply]
Ah, I should have been more clear - you don't need to use `spacy convert` at all. That command is only if you already have the old JSON format, or if you need to convert CoNLL data.

Normally you just save a DocBin to use as training data. Building on the previous example that would look like this:

```
from spacy.training import Example
from spacy.tokens import DocBin

TRAINING_DATA = ... # a list of your training data instances
db = DocBin()
for text, gold_dict in TRAINING_DATA:
    example = Example.from_dict(nlp.make_doc(text), gold_dict)
    db.add(example.reference)

db.to_disk("train.spacy")
```

You can see an example of this in action [in a sample project](https://github.com/explosion/projects/blob/3c17ba90490301e8665503a6516adf3f77bb5b07/pipelines/parser_demo/scripts/convert.py).

Does that make sense?

**Comment by dmsfabiano** [comment]
@polm I think the point here is that the documentation it's pretty lackluster. 

It's so indepth for 3.0 that a simple ""hello world"" guide was missed.

Do you have concrete examples on how to perform transfer learning while using any of the spacy default models (locally)?
